presto-main/src/main/java/com/facebook/presto/server/remotetask/WorkerTaskStatusFetcher.java

```diff
@@ -0,0 +1,283 @@
+                updateTaskStatus(value);
+                errorTracker.requestSucceeded();
+            }
+            finally {
+                scheduleNextRequest();
+            }
+        }
+    }
+
+    @Override
+    public void failed(Throwable cause)
+    {
+        try (SetThreadName ignored = new SetThreadName("ContinuousTaskStatusFetcher-%s", taskId)) {
+            updateStats(currentRequestStartNanos.get());
+            try {
+                // if task not already done, record error
+                TaskStatus taskStatus = getTaskStatus();
+                if (!taskStatus.getState().isDone()) {
+                    errorTracker.requestFailed(cause);
+                }
+            }
+            catch (Error e) {
+                onFail.accept(e);
+                throw e;
+            }
+            catch (RuntimeException e) {
+                onFail.accept(e);
+            }
+            finally {
+                scheduleNextRequest();
+            }
+        }
+    }
+
+    @Override
+    public void fatal(Throwable cause)
+    {
+        try (SetThreadName ignored = new SetThreadName("ContinuousTaskStatusFetcher-%s", taskId)) {
+            updateStats(currentRequestStartNanos.get());
+            onFail.accept(cause);
+        }
+    }
+
+    void updateTaskStatus(TaskStatus newValue)
+    {
+        // change to new value if old value is not changed and new value has a newer version
+        AtomicBoolean taskMismatch = new AtomicBoolean();
+        taskStatus.setIf(newValue, oldValue -> {
+            // did the task instance id change
+            boolean isEmpty = oldValue.getTaskInstanceIdLeastSignificantBits() == 0 && oldValue.getTaskInstanceIdMostSignificantBits() == 0;
+            if (!isEmpty &&
+                    !(oldValue.getTaskInstanceIdLeastSignificantBits() == newValue.getTaskInstanceIdLeastSignificantBits() &&
+                            oldValue.getTaskInstanceIdMostSignificantBits() == newValue.getTaskInstanceIdMostSignificantBits())) {
+                taskMismatch.set(true);
+                return false;
+            }
+
+            if (oldValue.getState().isDone()) {
+                // never update if the task has reached a terminal state
+                return false;
+            }
+            if (newValue.getVersion() < oldValue.getVersion()) {
+                // don't update to an older version (same version is ok)
+                return false;
+            }
+            return true;
+        });
+
+        if (taskMismatch.get()) {
+            // This will also set the task status to FAILED state directly.
+            // Additionally, this will issue a DELETE for the task to the worker.
+            // While sending the DELETE is not required, it is preferred because a task was created by the previous request.
+            onFail.accept(new PrestoException(REMOTE_TASK_MISMATCH, format("%s (%s)", REMOTE_TASK_MISMATCH_ERROR, HostAddress.fromUri(getTaskStatus().getSelf()))));
```

Comment:
we need to keep this also per task and call the one fothose which failed or had issue when fetching response.